### ISSUE TITLE
Doesn't compile with -DNDEBUG

### ISSUE THREAD

**Issue by Kicer86** [majn/telegram-purple#429]
Some examples of erros:

mtproto-client.c: In function ‘process_respq_answer’:
./auto.h:39:63: error: ‘tl_type_res_p_q’ undeclared (first use in this function); did you mean ‘tl_type_descr’?
 #define TYPE_TO_PARAM(NAME) (&(struct paramed_type) {.type = &tl_type_## NAME, .params=0})
                                                               ^
mtproto-client.c:362:22: note: in expansion of macro ‘TYPE_TO_PARAM’
   if (skip_type_any (TYPE_TO_PARAM (res_p_q)) < 0 || in_ptr != in_end) {
                      ^~~~~~~~~~~~~
./auto.h:39:63: note: each undeclared identifier is reported only once for each function it appears in
 #define TYPE_TO_PARAM(NAME) (&(struct paramed_type) {.type = &tl_type_## NAME, .params=0})
                                                               ^
mtproto-client.c:362:22: note: in expansion of macro ‘TYPE_TO_PARAM’
   if (skip_type_any (TYPE_TO_PARAM (res_p_q)) < 0 || in_ptr != in_end) {
                      ^~~~~~~~~~~~~
mtproto-client.c: In function ‘process_dh_answer’:
./auto.h:39:63: error: ‘tl_type_server_d_h_params’ undeclared (first use in this function); did you mean ‘bl_do_set_dh_params’?
 #define TYPE_TO_PARAM(NAME) (&(struct paramed_type) {.type = &tl_type_## NAME, .params=0})
                                                               ^

**Comment by BenWiederhake** [comment]
Oh, that's bad.  Is it possible that the submodules are out of sync, and you need to `git submodule sync && git submodule update` first?

Which commit/branch are you using?  I tried both `master` and `dev-1.4.0`, both build fine on my machine.  tgp is meant to be built using `configure && make`, did you use that or a different toolchain?

And if none of the above work:  What exactly is the gcc invocation that caused this?  And which version?  (I'm using 7.2.1 20171025)

**Comment by Kicer86** [comment]
I'm using Arch's AUR to build it. I'll have to make a closer look if you are not maintainer of AUR's package.

**Comment by Kicer86** [comment]
So I did a reaserch, and the problem was with my enviromental flags. I had -DNDEBUG in CFLAGS. Without this macro everything compiles. Yet it is still weird.

**Comment by majn** [comment]
Under no circumstance ever remove the assertions, those are used by libtgl to check properties of the cryptographic functions. Removing them would render the Diffie-Hellman implementation vulnerable to attacks.

**Comment by BenWiederhake** [comment]
tgp, tgl, and tl-parser assume that `assert` always evaluates all side-effects.  Proof:

```
tgl/tl-parser/tl-parser.c:      assert (write (__f, &x_c, 1) == 1);
tgl/mtproto-client.c:    assert (TLS->net_methods->write_out (c, &total_len, 1) == 1);
tgl/queries.c:    assert (write (D->fd, ptr, len) == len);
tgl/tgl-net.c:  assert (tgln_write_out (c, &byte, 1) == 1);
telegram-base.c:  assert (write (auth_file_fd, DC->auth_key, 256) == 256);
tgp-net.c:  assert (tgln_write_out (c, &byte, 1) == 1);
(and numerous other lines)
```

Another proof: Very early during build, `tgl/auto/scheme.tlo` ends up being *empty* because the writes to this file appear exclusively in `assert`s.

- Changing all of these individually is a huge effort, made pointless by the fact that you really really want those sanity-checks, in order to fail-early in case of bugs.
- Changing the definition of `assert` to always evaluate side-effects is stupid, because it doesn't improve run-time at all.

So the best I can do is insert a compile-time error at the top of the file, warning against defining `NDEBUG`.

**Comment by BenWiederhake** [comment]
Actually yeah, I'm going to do that.  I wrote a patch that has the following effect:

```
$ ./configure CFLAGS="-DNDEBUG"
checking for gcc... gcc
checking whether the C compiler works... yes
checking for C compiler default output file name... a.out
checking for suffix of executables... 
checking whether we are cross compiling... no
checking for suffix of object files... o
checking whether we are using the GNU C compiler... yes
checking whether gcc accepts -g... yes
checking for gcc option to accept ISO C89... none needed
checking for a BSD-compatible install... /usr/bin/install -c
configure: error: The macro NDEBUG cannot work.  All writes and reads are wrapped in asserts.
For more info, see https://github.com/majn/telegram-purple/issues/429
```

**Comment by BenWiederhake** [comment]
Should be fixed in [v1.4.1](https://github.com/majn/telegram-purple/releases/tag/v1.4.1).
